In [1]:
import pandas as pd
import collections
import gensim
from gensim import corpora
from gensim.models import CoherenceModel
from gensim.models import LdaModel


C:\Users\suzan\anaconda3\lib\site-packages\scipy\__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.24.3
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


In [2]:
# load the dataset
# df = pd.read_csv('ds_test/preprocessed_topicmoddeling_masterset.csv', low_memory=False, lineterminator='\n')
df = pd.read_csv('reddit_climatechange_dataset_sentiment-v2.csv', low_memory=False, lineterminator='\n')

In [3]:
df_dtm = df

### Document Term Matrix

In [4]:
#removing nulls
mask = df_dtm['body-tm'].notnull()
df_dtm = df_dtm[mask]

In [5]:
# convert the preprocessed text column to a list of strings
documents = list(df_dtm['body-tm'])

# tokenize the documents
tokenized_docs = [doc.split() for doc in documents]

# create a dictionary from the tokenized documents
dictionary = corpora.Dictionary(tokenized_docs)

# create a corpus using the dictionary and the tokenized documents
corpus = [dictionary.doc2bow(doc) for doc in tokenized_docs]

In [6]:
# Load the saved LDA model
lda_model = LdaModel.load("tm_seed4894454_topics30.model")

In [7]:
num_topics = 30
total_words = 10

## Save topic coherence

In [8]:
# Define the topic number for which you want to calculate the coherence score
topic_number = 6

In [9]:
import datetime
before = datetime.datetime.now()
print(before)

2023-05-18 15:49:47.825696


In [10]:
# Compute coherence score for the specific topic
coherence_model = CoherenceModel(
    model=lda_model,
    texts=tokenized_docs,
    dictionary=dictionary,
    coherence='c_v'
)

In [11]:
after = datetime.datetime.now()
print(after)

2023-05-18 15:49:48.590518


In [12]:
coherence_score = coherence_model.get_coherence_per_topic()[7]
print(coherence_score)

0.7800189652738909


In [13]:
after = datetime.datetime.now()
print(after)

2023-05-18 15:54:56.777285


In [ ]:
topic_cluster_words = []

topic_words = lda_model.show_topic(topic_number, topn=10)
topic_words = [word for word, _ in topic_words]
topic_cluster_words.append(", ".join(topic_words))
print(topic_cluster_words)

topic 0:
topic 1: 0.6207699918619637
topic 2: 0.44741884738484644
topic 3: 0.2749774347322928
topic 4: 
topic 5: 0.6936560900569879
topic 6: 
topic 7: 0.7800189652738909
topic 8: 0.5143029831846286
topic 9: 0.5690835473600127
topic 10: 0.5280982069450864
topic 11: 
topic 12: 
topic 13:
topic 14: 
topic 15: 
topic 16: 
topic 17: 
topic 18: 
topic 19: 
topic 20: 
topic 21: 
topic 22: 
topic 23: 
topic 24: 
topic 25: 
topic 26: 
topic 27:
topic 28: 
topic 29: 

### all at once

In [15]:
topic_cluster_words = []
coh_score = []

for i in range(0, 30):
    # Get the top words for the specific topic
    topic_words = lda_model.show_topic(i, topn=10)
    topic_words = [word for word, _ in topic_words]
    topic_cluster_words.append(", ".join(topic_words))

    
    coherence_score = coherence_model.get_coherence_per_topic()[i]
    coh_score.append(coherence_score)
    print('topic', i, '  ', coherence_score, '  ', topic_cluster_words)
    

topic 0    0.387276669207062    ['one, would, also, may, term, many, different, even, example, time']
topic 1    0.6207699918619637    ['one, would, also, may, term, many, different, even, example, time', 'article, study, scientist, research, data, source, report, read, science, link']
topic 2    0.44741884738484644    ['one, would, also, may, term, many, different, even, example, time', 'article, study, scientist, research, data, source, report, read, science, link', 'year, last, time, ago, day, every, million, next, new, decade']
topic 3    0.2749774347322928    ['one, would, also, may, term, many, different, even, example, time', 'article, study, scientist, research, data, source, report, read, science, link', 'year, last, time, ago, day, every, million, next, new, decade', 'space, reactor, 2020, safe, thanks, projection, existential, jet, invasion, private']
topic 4    0.2609927397054235    ['one, would, also, may, term, many, different, even, example, time', 'article, study, scien

topic 14    0.3425234693838918    ['one, would, also, may, term, many, different, even, example, time', 'article, study, scientist, research, data, source, report, read, science, link', 'year, last, time, ago, day, every, million, next, new, decade', 'space, reactor, 2020, safe, thanks, projection, existential, jet, invasion, private', 'welcome, capitalism, system, re, arguing, headline, capitalist, ignoring, socialist, socialism', 'energy, nuclear, fuel, oil, fossil, carbon, power, gas, coal, plant', 'tax, government, money, company, public, policy, pay, business, law, dollar', 'global, warming, ice, temperature, weather, heat, winter, cold, extreme, arctic', 'say, people, said, think, one, point, know, saying, want, wrong', 'water, area, fire, level, sea, land, due, region, drought, city', 'population, human, collapse, crisis, resource, threat, future, humanity, disaster, risk', 'fucking, fuck, shit, dont, bullshit, stupid, ‘, minister, thats, anti', 'good, evil, wonder, faith, weird

topic 20    0.5470896521931932    ['one, would, also, may, term, many, different, even, example, time', 'article, study, scientist, research, data, source, report, read, science, link', 'year, last, time, ago, day, every, million, next, new, decade', 'space, reactor, 2020, safe, thanks, projection, existential, jet, invasion, private', 'welcome, capitalism, system, re, arguing, headline, capitalist, ignoring, socialist, socialism', 'energy, nuclear, fuel, oil, fossil, carbon, power, gas, coal, plant', 'tax, government, money, company, public, policy, pay, business, law, dollar', 'global, warming, ice, temperature, weather, heat, winter, cold, extreme, arctic', 'say, people, said, think, one, point, know, saying, want, wrong', 'water, area, fire, level, sea, land, due, region, drought, city', 'population, human, collapse, crisis, resource, threat, future, humanity, disaster, risk', 'fucking, fuck, shit, dont, bullshit, stupid, ‘, minister, thats, anti', 'good, evil, wonder, faith, weird

topic 25    0.5861122337767515    ['one, would, also, may, term, many, different, even, example, time', 'article, study, scientist, research, data, source, report, read, science, link', 'year, last, time, ago, day, every, million, next, new, decade', 'space, reactor, 2020, safe, thanks, projection, existential, jet, invasion, private', 'welcome, capitalism, system, re, arguing, headline, capitalist, ignoring, socialist, socialism', 'energy, nuclear, fuel, oil, fossil, carbon, power, gas, coal, plant', 'tax, government, money, company, public, policy, pay, business, law, dollar', 'global, warming, ice, temperature, weather, heat, winter, cold, extreme, arctic', 'say, people, said, think, one, point, know, saying, want, wrong', 'water, area, fire, level, sea, land, due, region, drought, city', 'population, human, collapse, crisis, resource, threat, future, humanity, disaster, risk', 'fucking, fuck, shit, dont, bullshit, stupid, ‘, minister, thats, anti', 'good, evil, wonder, faith, weird

topic 29    0.7728428827056761    ['one, would, also, may, term, many, different, even, example, time', 'article, study, scientist, research, data, source, report, read, science, link', 'year, last, time, ago, day, every, million, next, new, decade', 'space, reactor, 2020, safe, thanks, projection, existential, jet, invasion, private', 'welcome, capitalism, system, re, arguing, headline, capitalist, ignoring, socialist, socialism', 'energy, nuclear, fuel, oil, fossil, carbon, power, gas, coal, plant', 'tax, government, money, company, public, policy, pay, business, law, dollar', 'global, warming, ice, temperature, weather, heat, winter, cold, extreme, arctic', 'say, people, said, think, one, point, know, saying, want, wrong', 'water, area, fire, level, sea, land, due, region, drought, city', 'population, human, collapse, crisis, resource, threat, future, humanity, disaster, risk', 'fucking, fuck, shit, dont, bullshit, stupid, ‘, minister, thats, anti', 'good, evil, wonder, faith, weird

In [16]:
results_df = pd.DataFrame({
    'topic_number': range(30),
    'coherence_score': coh_score
})

In [17]:
results_df

topic_number  coherence_score
0              0         0.387277
1              1         0.620770
2              2         0.447419
3              3         0.274977
4              4         0.260993
5              5         0.693656
6              6         0.621133
7              7         0.780019
8              8         0.514303
9              9         0.569084
10            10         0.528098
11            11         0.447932
12            12         0.315162
13            13         0.481930
14            14         0.342523
15            15         0.500843
16            16         0.428933
17            17         0.627592
18            18         0.805021
19            19         0.700112
20            20         0.547090
21            21         0.484818
22            22         0.582579
23            23         0.651227
24            24         0.422270
25            25         0.586112
26            26         0.294959
27            27         0.504235
28            28         0.482761
29            29         0.772843

In [18]:
results_df.to_csv('topic_coherence-scores.csv', index=False)